### ISSUE TITLE
Steam hangs on startup, potentially because of Mesa 23.1.4

### ISSUE THREAD

**Issue by CobaltCause** [NixOS/nixpkgs#248439]
### Describe the bug
Starting Steam results in the "logging in" window (the one after the updater progress bar but before the main window) working for a moment, then becoming black and then hanging. Even trying to close the window has no effect.

### Steps To Reproduce
1. Upgrade to ce5e4a6ef2e59d89a971bc434ca8ca222b9c7f5e (`nixos-unstable`)
2. Reboot (maybe not required?)
3. Run steam
4. It doesn't work

### Expected behavior
Steam should open its main window and be generally usable.

### Additional context

Downgrading Mesa from 23.1.4 to 23.1.3 (reboot maybe required?) makes the issue go away. Not sure if this is directly the cause of the problem though.

GPU: XFX RX 7900 XTX (more Xs => more gooder)

### Notify maintainers

Steam: @jagajaga, @jonringer 
Mesa: @primeos, @vcunat

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.9, NixOS, 23.11 (Tapir), 23.11.20230810.ce5e4a6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - nixpkgs: `/nix/store/63s5a8i5vsrf5wkmfzvk5hyiim24p2jr-source`
```

**Comment by CobaltCause** [comment]
I used the `hardware.opengl.package{,32}` options to try Mesa 23.1.5, which didn't fix it, but did make the screen not go black. I still get stuck in the same place though. Using that same mechanism to downgrade to Mesa 23.1.3 fixes it too. (Rebooting between every switch for good measure.)

**Comment by CobaltCause** [comment]
I get a black screen and hang on the "logging in" window on my laptop's i7-1185G7 iGPU on Mesa 23.1.4 too.

**Comment by Atemu** [comment]
Steam is working just fine for me on 23.1.4 with an RX6800XT. How did you install Steam?

**Comment by CobaltCause** [comment]
https://or.computer.surgery/charles/dotfiles/-/blob/c19da451505572c1b3aa3c020b317c8e74c5a611/nixos/conf/graphical/games.nix#L49

**Comment by CobaltCause** [comment]
On the 7900 XTX and Mesa 23.1.4, sometimes Steam gets past the "logging in" window to the main window, which is fully transparent. I can even interact with it, and the popups for when you click on e.g. the `Steam` text in the top left corner show up correctly. But after a few seconds it crashes (the window closes), and I see this in `dmesg`:

```
[  596.299229] llvmpipe-25[10313]: segfault at 7fcdf573dcfc ip 00007fcf139893aa sp 00007fcef65e8380 error 4 likely on CPU 8 (core 8, socket 0)
[  596.299235] Code: 62 b2 65 28 00 d9 62 01 7d 29 6f c6 62 91 0d 20 66 c8 62 02 3d 20 39 c6 62 62 3d 20 40 c4 62 91 3d 20 fe ea 62 f1 3d a1 fe e4 <c4> 62 35 90 04 28 c5 d1 ef ed 62 91 5d 28 fe f2 62 91 25 20 fe f8
```

And this in `journalctl --user -f`:

<details>
<summary>Click to show because it's long</summary>

```
Aug 11 15:43:16 compy systemd-coredump[10570]: [🡕] Process 9881 (steamwebhelper) of user 1000 dumped core.

                                               Module /nix/store/qbd4avam8s7mhjnabf2dwh0xsc7wqa3l-libpciaccess-0.17/lib/libpciaccess.so.0.11.1 without build-id.
                                               Module /nix/store/1ql93amj3rmq5yg7qry7ky6gwpbvdgxp-ncurses-6.4/lib/libncursesw.so.6.4 without build-id.
                                               Module /nix/store/qx2y7chays0azsb9zfk2qx7z3hjzmvhl-libdrm-2.4.115/lib/libdrm_intel.so.1.0.0 without build-id.
                                               Module /nix/store/qx2y7chays0azsb9zfk2qx7z3hjzmvhl-libdrm-2.4.115/lib/libdrm_nouveau.so.2.0.0 without build-id.
                                               Module /nix/store/qx2y7chays0azsb9zfk2qx7z3hjzmvhl-libdrm-2.4.115/lib/libdrm_amdgpu.so.1.0.0 without build-id.
                                               Module /nix/store/6asa875111lg4prkfmbkazwk2i5qvabb-libelf-0.8.13/lib/libelf.so.0.8.13 without build-id.
                                               Module /nix/store/qx2y7chays0azsb9zfk2qx7z3hjzmvhl-libdrm-2.4.115/lib/libdrm_radeon.so.1.0.1 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-xfixes.so.0.0.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-sync.so.1.0.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-present.so.0.0.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-randr.so.0.1.0 without build-id.
                                               Module /nix/store/6wxyydv172rrpbbv9vds1nllbnb9dw8b-libxshmfence-1.3.2/lib/libxshmfence.so.1.0.0 without build-id.
                                               Module /nix/store/jp1gws2r3lvhsrv2milpan972m0gz51w-libXxf86vm-1.1.5/lib/libXxf86vm.so.1.0.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-dri2.so.0.0.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-glx.so.0.0.0 without build-id.
                                               Module /nix/store/pm4gxdsl1jjn3yr063si78b6636vvda7-systemd-253.6/lib/libudev.so.1.7.6 without build-id.
                                               Module /nix/store/81crz17ih5y188f0r30w5cbcagjgf7l0-libXScrnSaver-1.2.4/lib/libXss.so.1.0.0 without build-id.
                                               Module /nix/store/vcchv1mm397wi4cx2g1dk0793d7a77pn-libgpg-error-1.47/lib/libgpg-error.so.0.34.0 without build-id.
                                               Module /nix/store/mrb32818q23q7b8pjvnq1fdrwdmmbkfy-freetype-2.13.1/lib/libfreetype.so.6.20.0 without build-id.
                                               Module /nix/store/v5llm7rmipvg487yvgb15q0m2jxrihq3-libidn2-2.3.4/lib/libidn2.so.0.3.8 without build-id.
                                               Module /nix/store/nglndag1zm1af90s6ab172zpa44bd24i-libgcrypt-1.10.2/lib/libgcrypt.so.20.4.2 without build-id.
                                               Module /nix/store/4igdc32rmnijcra8y3r1h42987ghzag2-gcc-12.3.0-lib/lib/libssp.so.0.0.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-render.so.0.0.0 without build-id.
                                               Module /nix/store/i1wjf5x2jdzbyxyqrywgbzbdcsvaljv2-libglvnd-1.6.0/lib/libEGL.so.1.1.0 without build-id.
                                               Module /nix/store/mppz0yhlnna153fmw03hd4r364mq07z6-libpng-apng-1.6.39/lib/libpng16.so.16.39.0 without build-id.
                                               Module /nix/store/nqr0hwhybhx7jz6lz1xhazl4d9fvf03c-bzip2-1.0.8/lib/libbz2.so.1.0.8 without build-id.
                                               Module /nix/store/hysm64xr80n6ykrlqrsqwbxv12ihn715-harfbuzz-7.3.0/lib/libharfbuzz.so.0.60730.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-shm.so.0.0.0 without build-id.
                                               Module /nix/store/zrhgwwrdipfmjy3kawr0sf93y2f0x0qs-xz-5.4.3/lib/liblzma.so.5.4.3 without build-id.
                                               Module /nix/store/d729a24im9pcbz5hmdvvri4brwqby8kk-libcap-2.69-lib/lib/libcap.so.2.69 without build-id.
                                               Module /nix/store/jlkh099qcdd2j919h2y217w3xgfss6q0-fontconfig-2.14.2-lib/lib/libfontconfig.so.1.12.0 without build-id.
                                               Module /nix/store/pm4gxdsl1jjn3yr063si78b6636vvda7-systemd-253.6/lib/libsystemd.so.0.36.0 without build-id.
                                               Module /nix/store/pbgggi1xlblgwhdshqawqklj7wiknycs-libthai-0.1.29/lib/libthai.so.0.3.1 without build-id.
                                               Module /nix/store/ah31a45w9n6r45idnrxg24gy0pkm5r0f-pango-1.50.14/lib/libpangoft2-1.0.so.0.5000.14 without build-id.
                                               Module /nix/store/5jrnzkg39q8jrjj7qnkqijs4ixqda8d2-nspr-4.35/lib/libplc4.so without build-id.
                                               Module /nix/store/5jrnzkg39q8jrjj7qnkqijs4ixqda8d2-nspr-4.35/lib/libplds4.so without build-id.
                                               Module /nix/store/qx2y7chays0azsb9zfk2qx7z3hjzmvhl-libdrm-2.4.115/lib/libdrm.so.2.4.0 without build-id.
                                               Module /nix/store/i1wjf5x2jdzbyxyqrywgbzbdcsvaljv2-libglvnd-1.6.0/lib/libGLdispatch.so.0.0.0 without build-id.
                                               Module /nix/store/i1wjf5x2jdzbyxyqrywgbzbdcsvaljv2-libglvnd-1.6.0/lib/libGLX.so.0.0.0 without build-id.
                                               Module /nix/store/p9a2nhhpa2dwyw1sy5gr4482ddqmwpkx-zlib-1.2.13/lib/libz.so.1.2.13 without build-id.
                                               Module /nix/store/kbjir9mdrz2w1f7kl4ikisg491x2xd1y-gcc-12.3.0-libgcc/lib/libgcc_s.so.1 without build-id.
                                               Module /nix/store/sal68rv7jr1jiicvaxgc5zsjdgz6bk2n-at-spi2-core-2.48.3/lib/libatspi.so.0.0.1 without build-id.
                                               Module /nix/store/1p54jn98w5s74y818csmn9955p3f8q4y-cups-2.4.6-lib/lib/libcups.so.2 without build-id.
                                               Module /nix/store/ah31a45w9n6r45idnrxg24gy0pkm5r0f-pango-1.50.14/lib/libpango-1.0.so.0.5000.14 without build-id.
                                               Module /nix/store/ah31a45w9n6r45idnrxg24gy0pkm5r0f-pango-1.50.14/lib/libpangocairo-1.0.so.0.5000.14 without build-id.
                                               Module /nix/store/qd8js1fx8r4y465kni2n0nbyzl6drj5f-alsa-lib-1.2.9/lib/libasound.so.2.0.0 without build-id.
                                               Module /nix/store/0dsl1d89lq5ljq4xz49c06m2h8qaa53q-dbus-1.14.8-lib/lib/libdbus-1.so.3.32.3 without build-id.
                                               Module /nix/store/cnlf5ww1hgqhynq1g2f73k4ahbdpav7v-expat-2.5.0/lib/libexpat.so.1.8.10 without build-id.
                                               Module /nix/store/qcyayd6skll6143jwsxraq1yy25jcikx-libXfixes-6.0.1/lib/libXfixes.so.3.1.0 without build-id.
                                               Module /nix/store/jh4jv0v6pxjdh7zcrg4vqg0j1c19vwmh-libXcursor-1.2.1/lib/libXcursor.so.1.0.2 without build-id.
                                               Module /nix/store/sal68rv7jr1jiicvaxgc5zsjdgz6bk2n-at-spi2-core-2.48.3/lib/libatk-bridge-2.0.so.0.0.0 without build-id.
                                               Module /nix/store/4igdc32rmnijcra8y3r1h42987ghzag2-gcc-12.3.0-lib/lib/libstdc++.so.6.0.30 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb-dri3.so.0.1.0 without build-id.
                                               Module /nix/store/g93jp1k5j47zxqjnd1h0a8aixhiczz28-libxcb-1.15/lib/libxcb.so.1.1.0 without build-id.
                                               Module /nix/store/sal68rv7jr1jiicvaxgc5zsjdgz6bk2n-at-spi2-core-2.48.3/lib/libatk-1.0.so.0.24812.1 without build-id.
                                               Module /nix/store/5jrnzkg39q8jrjj7qnkqijs4ixqda8d2-nspr-4.35/lib/libnspr4.so without build-id.
                                               Module /nix/store/dlqai2l8phhxxrkr2fydh5r1f0azf7h1-nss-3.90/lib/libnss3.so without build-id.
                                               Module /nix/store/k3w1agqdhvqpzzia68vkjwj71qdz93jw-libX11-1.8.6/lib/libX11-xcb.so.1.0.0 without build-id.
                                               Module /nix/store/dlqai2l8phhxxrkr2fydh5r1f0azf7h1-nss-3.90/lib/libsmime3.so without build-id.
                                               Module /nix/store/dlqai2l8phhxxrkr2fydh5r1f0azf7h1-nss-3.90/lib/libnssutil3.so without build-id.
                                               Module /nix/store/i1wjf5x2jdzbyxyqrywgbzbdcsvaljv2-libglvnd-1.6.0/lib/libGL.so.1.7.0 without build-id.
                                               Module /nix/store/yyar0v0qlwbnlvjhh1vssp5bnf649nhm-libXdamage-1.1.6/lib/libXdamage.so.1.1.0 without build-id.
                                               Module /nix/store/zi9rg9a4vjqky3gia3ck9421bs88cf6z-libXcomposite-0.4.6/lib/libXcomposite.so.1.0.0 without build-id.
                                               Module /nix/store/k3w1agqdhvqpzzia68vkjwj71qdz93jw-libX11-1.8.6/lib/libX11.so.6.4.0 without build-id.
                                               Module /nix/store/y9amcxznwjajhp8m01qdq9bd8vq6i76v-libXrandr-1.5.3/lib/libXrandr.so.2.2.0 without build-id.
                                               Module /nix/store/ny6j15w7ibhrym60dhqgik475137fwsk-libXtst-1.2.4/lib/libXtst.so.6.1.0 without build-id.
                                               Module /nix/store/2yn2dy4c7a2rsjrc5s7anllwi7c29760-libXrender-0.9.11/lib/libXrender.so.1.3.0 without build-id.
                                               Module /nix/store/hz2cw1szf2xmy8lcryz685qi50l58pz3-libXext-1.3.5/lib/libXext.so.6.4.0 without build-id.
                                               Module /nix/store/04p7x1xhrc6hiws0gwpnvhyls5dq9hsz-libXi-1.8.1/lib/libXi.so.6.1.0 without build-id.
                                               Module libsensors.so.5 without build-id.
                                               Module libbrotlicommon.so.1 without build-id.
                                               Module libgmp.so.10 without build-id.
                                               Module libhogweed.so.6 without build-id.
                                               Module libnettle.so.8 without build-id.
                                               Module libtasn1.so.6 without build-id.
                                               Module libunistring.so.5 without build-id.
                                               Module libp11-kit.so.0 without build-id.
                                               Module libbrotlidec.so.1 without build-id.
                                               Module liblz4.so.1 without build-id.
                                               Module libzstd.so.1 without build-id.
                                               Module libgnutls.so.30 without build-id.
                                               Module libavahi-client.so.3 without build-id.
                                               Module libavahi-common.so.3 without build-id.
                                               Module libfribidi.so.0 without build-id.
                                               Module libpcre2-8.so.0 without build-id.
                                               Module libffi.so.8 without build-id.
                                               Stack trace of thread 10313:
                                               #0  0x00007fcf139893aa n/a (n/a + 0x0)
                                               #1  0x00007fcf014f6a9b lp_rast_shade_tile (swrast_dri.so + 0x6f6a9b)
                                               #2  0x00007fcf014f7045 rasterize_scene (swrast_dri.so + 0x6f7045)
                                               #3  0x00007fcf014f728c thread_function (swrast_dri.so + 0x6f728c)
                                               #4  0x00007fcf00f0f0e7 impl_thrd_routine (swrast_dri.so + 0x10f0e7)
                                               #5  0x00007fcf2ca9fdd4 start_thread (/nix/store/ibp4camsx1mlllwzh32yyqcq2r2xsy1a-glibc-2.37-8/lib/libc.so.6 + 0x85dd4)
                                               #6  0x00007fcf2cb219b0 __clone3 (/nix/store/ibp4camsx1mlllwzh32yyqcq2r2xsy1a-glibc-2.37-8/lib/libc.so.6 + 0x1079b0)

                                               Stack trace of thread 9881:
                                               #0  0x00007fcf2cb1477f __poll (/nix/store/ibp4camsx1mlllwzh32yyqcq2r2xsy1a-glibc-2.37-8/lib/libc.so.6 + 0xfa77f)
                                               #1  0x00007fcf2d7330ce g_main_context_iterate.constprop.0 (/nix/store/5gs8a87g7dfy1pgylzc7bgri68fd8a66-glib-2.76.4/lib/libglib-2.0.so.0.7600.4 + 0x5e0ce)
                                               #2  0x00007fcf2d7331ec g_main_context_iteration (/nix/store/5gs8a87g7dfy1pgylzc7bgri68fd8a66-glib-2.76.4/lib/libglib-2.0.so.0.7600.4 + 0x5e1ec)
                                               #3  0x00007fcf31ddfa72 n/a (/home/charles/.local/share/Steam/ubuntu12_64/libcef.so + 0x43dfa72)
                                               #4  0x00007fcf31e272a4 n/a (/home/charles/.local/share/Steam/ubuntu12_64/libcef.so + 0x44272a4)
                                               #5  0x00007fcf31dfee5e n/a (/home/charles/.local/share/Steam/ubuntu12_64/libcef.so + 0x43fee5e)
                                               #6  0x00007fcf31cfb753 n/a (/home/charles/.local/share/Steam/ubuntu12_64/libcef.so + 0x42fb753)
                                               #7  0x0000000000527e83 n/a (/home/charles/.local/share/Steam/ubuntu12_64/steamwebhelper + 0x127e83)
                                               ELF object binary architecture: AMD x86-64
```

</details>

**Comment by CobaltCause** [comment]
There are now a handful of upstream issues for what seems to be this same problem:

* https://github.com/ValveSoftware/steam-for-linux/issues/9974
* https://github.com/ValveSoftware/steam-for-linux/issues/9976
* https://github.com/ValveSoftware/steam-for-linux/issues/9977
* https://github.com/ValveSoftware/steam-for-linux/issues/9978

https://github.com/ValveSoftware/steam-for-linux/issues/9977 is particularly interesting because it lines up with the mention of llvmpipe in the above `dmesg` output I just encountered. In fact:

```
$ nix shell nixpkgs#mesa-demos -c glxinfo | rg '(OpenGL version string:|Device:)'
    Device: llvmpipe (LLVM 15.0.7, 256 bits) (0xffffffff)
OpenGL version string: 4.5 (Compatibility Profile) Mesa 23.1.4
$ nix shell nixpkgs#pkgsi686Linux.mesa-demos -c glxinfo | rg '(OpenGL version string:|Device:)'
    Device: llvmpipe (LLVM 15.0.7, 256 bits) (0xffffffff)
OpenGL version string: 4.5 (Compatibility Profile) Mesa 23.1.4
```

This does not look right.

**Comment by CobaltCause** [comment]
Ah, this problem with Mesa causing llvmpipe to be used is triggering this other, older bug in Steam: https://github.com/ValveSoftware/steam-for-linux/issues/9752. There's an open issue on Mesa for this problem as well: https://gitlab.freedesktop.org/mesa/mesa/-/issues/9345. It says that this is a bug in Steam because of a bug in SDL, which is here: https://github.com/libsdl-org/SDL/issues/7975.

So to summarize, there are 2 distinct problems here:

1. Mesa is broken in a way that causes llvmpipe to always be used
2. SDL and thus Steam is broken in a way that causes rendering on llvmpipe to not work

**Comment by CobaltCause** [comment]
I've opened an upstream issue for this since it seems like users on Arch and Ubuntu are also having this problem based on the Steam issues linked above: https://gitlab.freedesktop.org/mesa/mesa/-/issues/9527.

**Comment by Atemu** [comment]
I really don't think this is a mesa issue tbh. I even played VR with 23.1.4 yesterday and it was working fine.

One factor that might be different is that I should still be running the Steam runtime beta in order to catch new pressure-vessel issues on NixOS early.

**Comment by CobaltCause** [comment]
But how could

> ```
> $ nix shell nixpkgs#mesa-demos -c glxinfo | rg '(OpenGL version string:|Device:)'
>     Device: llvmpipe (LLVM 15.0.7, 256 bits) (0xffffffff)
> OpenGL version string: 4.5 (Compatibility Profile) Mesa 23.1.4
> $ nix shell nixpkgs#pkgsi686Linux.mesa-demos -c glxinfo | rg '(OpenGL version string:|Device:)'
>     Device: llvmpipe (LLVM 15.0.7, 256 bits) (0xffffffff)
> OpenGL version string: 4.5 (Compatibility Profile) Mesa 23.1.4
> ```

&mdash; https://github.com/NixOS/nixpkgs/issues/248439#issuecomment-1675509492

not be a Mesa issue?

**Comment by CobaltCause** [comment]
The way it would not be a Mesa issue is apparently if my compositor was built against Mesa 23.1.3 and every other program was built against Mesa 23.1.4. This is what I get for using flakes I guess. The solution was `inputs.hyprland.inputs.nixpkgs.follows = "nixpkgs";` to make sure the compositor is built against the same version of Mesa as everything else.
